django_lightweight_queue/backends/redis.py

```diff
@@ -20,6 +25,15 @@ def enqueue(self, job, queue):
         self.client.lpush(self._key(queue), job.to_json().encode('utf-8'))
 
     def dequeue(self, queue, worker_num, timeout):
+        if self.is_paused(queue):
+            # Block for a while to avoid constant polling ...
+            block_for_time(
+                lambda: self.is_paused(queue),
+                timeout=datetime.timedelta(minutes=5),
+            )
+            # ... but always indicate that we did no work
+            return None
```

Comment:
Just to make sure I understand this as this isn't super clear on first read:

- We enter the job loop (dequeue).
- We detect the queue is paused.
- We lock the worker for up to 5 minutes
- If we detect in the meantime that we are unpaused we bail out, releasing the job back into the queue to be picked up immediately by another worker.
- If times up we bail out, releasing the job back into the queue to be picked up later.

I think that second to last point is the one that could be confusing. Why would we not just carry on with the job in that case? Are there external timeout to consider? 

Also could this create issues with serialisation, i.e. for single worker queues where we rely on this for serialisation, wouldn't releasing the job back put it at the end of the queue?

Not pushing against it, but might be worth clarifying what we expect to happen in a comment, e.g. why is it important that _we did no work_.

Reply:
> * If we detect in the meantime that we are unpaused we bail out, releasing the job back into the queue to be picked up immediately by another worker.

I don't think this is what happens here. My intent and having re-read the code I believe what happens, was that in the case that the queue is paused we don't actually pull any work from redis and then always return "nothing-happened" (after a suitable delay).

I agree the behaviour you describe would be confusing, though I don't think that's what's happening here?

With regards to serial queues: only the `ReliableRedisBackend` (rather than this `RedisBackend`) has a mechanism to recover jobs from the processing keys. When it does so I believe ordering is preserved (as best it can) as it puts them back on the front of the queue.

My intent with returning without checking whether there was work to do was more about ensuring that the timing values recorded by the caller (namely `Worker.run` via `Worker.process` calls `dequeue`) were relatively accurate. Specifically to avoid real durations being exaggerated by the pauses. It also felt a bit more correct to do it this way.

> Are there external timeout to consider?

None that I had in mind, however I'd overlooked that the Worker class does already pass through a timeout to the backend, which we probably ought to respect. 8a263c8 fixes that.